pkg/solana/txm/pendingtx.go

```diff
@@ -3,78 +3,122 @@ package txm
 import (
 	"context"
 	"errors"
+	"fmt"
 	"sync"
 	"time"
 
 	"github.com/gagliardetto/solana-go"
-	"github.com/google/uuid"
 	"golang.org/x/exp/maps"
 )
 
+var ErrAlreadyInExpectedState = errors.New("transaction already in expected state")
+
 type PendingTxContext interface {
-	New(sig solana.Signature, cancel context.CancelFunc) (uuid.UUID, error)
-	Add(id uuid.UUID, sig solana.Signature) error
-	Remove(sig solana.Signature) uuid.UUID
+	// New adds a new tranasction in Broadcasted state to the storage
+	New(msg pendingTx, sig solana.Signature, cancel context.CancelFunc) error
+	// AddSignature adds a new signature for an existing transaction ID
+	AddSignature(id string, sig solana.Signature) error
+	// Remove removes transaction and related signatures from storage if not in finalized or errored state
+	Remove(sig solana.Signature) string
+	// ListAll returns all of the signatures being tracked for all transactions not yet finalized or errored
 	ListAll() []solana.Signature
-	Expired(sig solana.Signature, lifespan time.Duration) bool
-	// state change hooks
-	OnSuccess(sig solana.Signature) uuid.UUID
-	OnError(sig solana.Signature, errType int) uuid.UUID // match err type using enum
+	// Expired returns whether or not confirmation timeout amount of time has passed since creation
+	Expired(sig solana.Signature, confirmationTimeout time.Duration) bool
+	// OnProcessed marks transactions as Processed
+	OnProcessed(sig solana.Signature) (string, error)
+	// OnConfirmed marks transaction as Confirmed and moves it from broadcast map to confirmed map
+	OnConfirmed(sig solana.Signature) (string, error)
+	// OnFinalized marks transaction as Finalized, moves it from confirmed map to finalized map, removes signatures from signature map to stop confirmation checks
+	OnFinalized(sig solana.Signature, retentionTimeout time.Duration) (string, error)
+	// OnError marks transaction as errored, matches err type using enum, moves it from confirmed map to finalized map, removes signatures from signature map to stop confirmation checks
+	OnError(sig solana.Signature, retentionTimeout time.Duration, errType int) string
```

Comment:
why not returning an error here?

Reply:
Good point, added an error for consistency